src/Kestrel.Core/Internal/HttpConnectionMiddleware.cs

```diff
@@ -63,27 +63,20 @@ public Task OnConnectionAsync(ConnectionContext connectionContext)
 
             var connection = new FrameConnection(frameConnectionContext);
 
-            // The order here is important, start request processing so that
-            // the frame is created before this yields. Events need to be wired up
-            // afterwards
             var processingTask = connection.StartRequestProcessing(_application);
 
-            // Wire up the events an forward calls to the frame connection
-            // It's important that these execute synchronously because graceful
-            // connection close is order sensative (for now)
-            connectionContext.ConnectionAborted.ContinueWith((task, state) =>
+            // Abort the frame when the transport writer completes
+            connectionContext.Transport.Input.OnWriterCompleted((error, state) =>
```

Comment:
@pakrym I think this is rooting the FrameConnection causing the test to fail. We never remove the callbacks it seems.

Reply:
hmm, actually it looks like we do clear them...